mcs/tools/mkbundle/mkbundle.cs

```diff
@@ -1446,4 +1391,1193 @@ static string LocateFile(string default_path)
+
+				installedWindowsSDK = windowsSDK;
+			}
+
+			return installedWindowsSDK;
+		}
+
+		string FindCRuntimeSDKIncludePath (InstalledSDKInfo sdk)
+		{
+			string cRuntimeIncludePath = Path.Combine (sdk.InstallationFolder, "include");
+			if (sdk.IsSubVersion)
+				cRuntimeIncludePath = Path.Combine (cRuntimeIncludePath, sdk.Version.ToString ());
+
+			cRuntimeIncludePath = Path.Combine (cRuntimeIncludePath, "ucrt");
+			if (!Directory.Exists (cRuntimeIncludePath))
+				cRuntimeIncludePath = "";
+
+			return cRuntimeIncludePath;
+		}
+
+		string FindCRuntimeSDKLibPath (InstalledSDKInfo sdk)
+		{
+			string cRuntimeLibPath = Path.Combine (sdk.InstallationFolder, "lib");
+			if (sdk.IsSubVersion)
+				cRuntimeLibPath = Path.Combine (cRuntimeLibPath, sdk.Version.ToString ());
+
+			cRuntimeLibPath = Path.Combine (cRuntimeLibPath, "ucrt", Target64BitApplication () ? "x64" : "x86");
+			if (!Directory.Exists (cRuntimeLibPath))
+				cRuntimeLibPath = "";
+
+			return cRuntimeLibPath;
+		}
+
+		InstalledSDKInfo GetInstalledCRuntimeSDK ()
+		{
+			if (installedCRuntimeSDK == null) {
+				InstalledSDKInfo cRuntimeSDK = null;
+				var windowsSDK = GetInstalledWindowsSDK ();
+				var cRuntimeSDKs = GetInstalledCRuntimeSDKs ();
+
+				if (windowsSDK != null && cRuntimeSDKs != null) {
+					cRuntimeSDK = cRuntimeSDKs.Find (x => x.Version.ToString () == windowsSDK.Version.ToString ());
+					if (cRuntimeSDK == null && cRuntimeSDKs.Count != 0)
+						cRuntimeSDK = cRuntimeSDKs.First ();
+
+					installedCRuntimeSDK = cRuntimeSDK;
+				}
+			}
+
+			return installedCRuntimeSDK;
+		}
+
+		public void AddWindowsSDKIncludePaths (List<string> includePaths)
+		{
+			InstalledSDKInfo winSDK = GetInstalledWindowsSDK ();
+			if (winSDK != null) {
+				string winSDKIncludeDir = Path.Combine (winSDK.InstallationFolder, "include");
+
+				if (winSDK.IsSubVersion)
+					winSDKIncludeDir = Path.Combine (winSDKIncludeDir, winSDK.Version.ToString ());
+
+				// Include sub folders.
+				if (Directory.Exists (winSDKIncludeDir)) {
+					includePaths.Add (Path.Combine (winSDKIncludeDir, "um"));
+					includePaths.Add (Path.Combine (winSDKIncludeDir, "shared"));
+					includePaths.Add (Path.Combine (winSDKIncludeDir, "winrt"));
+				}
+			}
+
+			return;
+		}
+
+		public void AddWindowsSDKLibPaths (List<string> libPaths)
+		{
+			InstalledSDKInfo winSDK = GetInstalledWindowsSDK ();
+			if (winSDK != null) {
+				string winSDKLibDir = Path.Combine (winSDK.InstallationFolder, "lib");
+
+				if (winSDK.IsSubVersion) {
+					winSDKLibDir = Path.Combine (winSDKLibDir, winSDK.Version.ToString ());
+				} else {
+					// Older WinSDK's header folders are not versioned, but installed libraries are, use latest availalble version for now.
+					var winSDKLibDirInfo = new DirectoryInfo (winSDKLibDir);
+					var versions = winSDKLibDirInfo.GetDirectories ("*.*", SearchOption.TopDirectoryOnly).OrderByDescending (p => p.Name, new StringVersionComparer ());
```

Comment:
Can throw NRE on OrderByDescending

Reply:
Are you referring to GetDirectories returning null that could cause a NRE in above statement?

Reply:
yes

Reply:
Looking at the docs it says it will return empty array if no subdirectories are available, or throw exception if directory doesn't exist or other error conditions occur. If have tried above with empty directory scenario and the OrderByDescending doesn't cause an NRE, but next check on line 1762 is not correct when empty array, raising an InvalidOperationException. So fix is then to make sure that we have a items in versions (no empty collection) before doing the OrderByDescending and call to First() on collection.  OK?

Reply:
yep or you can call FirstOrDefault on it

Reply:
yep, will do.